src/validation.cpp

```diff
@@ -1029,46 +1041,86 @@ bool MemPoolAccept::Finalize(ATMPArgs& args, Workspace& ws)
 {
     AssertLockHeld(cs_main);
     LOCK(m_pool.cs); // mempool "read lock" (held through GetMainSignals().TransactionAddedToMempool())
 
     Workspace workspace(ptx);
 
-    if (!PreChecks(args, workspace)) return false;
+    if (!PreChecks(args, workspace)) return MempoolAcceptResult(workspace.m_state, workspace.m_ptx);
 
     // Only compute the precomputed transaction data if we need to verify
     // scripts (ie, other policy checks pass). We perform the inexpensive
     // checks first and avoid hashing and signature verification unless those
     // checks pass, to mitigate CPU exhaustion denial-of-service attacks.
     PrecomputedTransactionData txdata;
 
-    if (!PolicyScriptChecks(args, workspace, txdata)) return false;
+    if (!PolicyScriptChecks(args, workspace, txdata)) return MempoolAcceptResult(workspace.m_state, workspace.m_ptx);
 
-    if (!ConsensusScriptChecks(args, workspace, txdata)) return false;
+    if (!ConsensusScriptChecks(args, workspace, txdata)) return MempoolAcceptResult(workspace.m_state, workspace.m_ptx);
 
     // Tx was accepted, but not added
-    if (args.m_test_accept) return true;
+    if (args.m_test_accept) {
+        return MempoolAcceptResult(workspace.m_state, std::move(workspace.m_replaced_transactions), workspace.m_base_fees);
+    }
 
-    if (!Finalize(args, workspace)) return false;
+    if (!Finalize(args, workspace)) return MempoolAcceptResult(workspace.m_state, workspace.m_ptx);
 
     GetMainSignals().TransactionAddedToMempool(ptx, m_pool.GetAndIncrementSequence());
 
-    return true;
+    return MempoolAcceptResult(workspace.m_state, std::move(workspace.m_replaced_transactions), workspace.m_base_fees);
+}
+
+std::vector<MempoolAcceptResult> MemPoolAccept::AcceptMultipleTransactions(std::vector<CTransactionRef>& txns, ATMPArgs& args)
+{
+    AssertLockHeld(cs_main);
+    std::vector<Workspace> workspaces{};
+    workspaces.reserve(txns.size());
+    std::transform(txns.begin(), txns.end(), std::back_inserter(workspaces), [](CTransactionRef& tx) {
+        return Workspace(tx);
+    });
+
+    LOCK(m_pool.cs);
+    // Do all PreChecks first and fail fast to avoid running expensive script
+    // checks when unnecessary.
+    for (unsigned int i = 0; i < txns.size(); ++i) {
+        Workspace& workspace = workspaces[i];
+        if (!PreChecks(args, workspace)) {
+            return std::vector<MempoolAcceptResult> { MempoolAcceptResult(workspace.m_state, workspace.m_ptx) };
+        }
+        m_viewmempool.AddPackageTransaction(txns[i]);
+    }
+
+    // TODO: Enforce package-level feerate and other policies before script checks.
+    for (unsigned int i = 0; i < txns.size(); ++i) {
+        PrecomputedTransactionData txdata;
+        Workspace& workspace = workspaces[i];
+
+        if (!PolicyScriptChecks(args, workspace, txdata)) {
+            return std::vector<MempoolAcceptResult> { MempoolAcceptResult(workspace.m_state, workspace.m_ptx) };
+        }
+    }
```

Comment:
Consider using range based loops here:

```suggestion
    for (Workspace& ws : workspaces) {
        if (!PreChecks(args, ws)) {
            return std::vector<MempoolAcceptResult> { MempoolAcceptResult(ws.m_state, ws.m_ptx) };
        }
        m_viewmempool.AddPackageTransaction(ws.m_ptx);
    }

    // TODO: Enforce package-level feerate and other policies before script checks.
    for (Workspace& ws : workspaces) {
        PrecomputedTransactionData txdata;

        if (!PolicyScriptChecks(args, ws, txdata)) {
            return std::vector<MempoolAcceptResult> { MempoolAcceptResult(ws.m_state, ws.m_ptx) };
        }
    }
```